File: backend/internal/email/email.go
// Package email sends portfolio contact-form notifications via Amazon SES.
package email

import (
	"context"
	"fmt"

	"github.com/aws/aws-sdk-go-v2/aws"
	ses "github.com/aws/aws-sdk-go-v2/service/ses"
	"github.com/aws/aws-sdk-go-v2/service/ses/types"
)

// SESAPI is the narrow SES interface required by SendContactEmail.
// The real *ses.Client satisfies it.
type SESAPI interface {
	SendEmail(ctx context.Context, params *ses.SendEmailInput, optFns ...func(*ses.Options)) (*ses.SendEmailOutput, error)
}

// ContactEmailArgs holds the data needed to build and send the contact email.
type ContactEmailArgs struct {
	From      string
	To        string
	ReplyTo   string
	Name      string
	Email     string
	Message   string
	IP        string
	CreatedAt string
}

// SendContactEmail composes and dispatches a single SES email for a contact
// form submission. It mirrors the void TypeScript original: the SES output is
// discarded and only the error is returned.
func SendContactEmail(ctx context.Context, client SESAPI, a ContactEmailArgs) error {
	subject := "Portfolio: new contact from " + a.Name
	body := fmt.Sprintf("From: %s <%s>\nWhen: %s\nIP:   %s\n\n%s\n", a.Name, a.Email, a.CreatedAt, a.IP, a.Message)

	_, err := client.SendEmail(ctx, &ses.SendEmailInput{
		Source: aws.String(a.From),
		Destination: &types.Destination{
			ToAddresses: []string{a.To},
		},
		ReplyToAddresses: []string{a.ReplyTo},
		Message: &types.Message{
			Subject: &types.Content{
				Data: aws.String(subject),
			},
			Body: &types.Body{
				Text: &types.Content{
					Data: aws.String(body),
				},
			},
		},
	})
	return err
}

File: backend/src/ingest/email.ts
// backend/src/ingest/email.ts
import { SESClient, SendEmailCommand } from '@aws-sdk/client-ses';

export interface ContactEmailArgs {
  from: string;       // verified SES identity, e.g. noreply@trystan-tbm.dev
  to: string;         // recipient (from Secrets Manager)
  replyTo: string;    // submitter's email, so a reply goes straight to them
  name: string;
  email: string;
  message: string;
  ip: string;
  createdAt: string;  // ISO-8601
}

export async function sendContactEmail(ses: SESClient, a: ContactEmailArgs): Promise<void> {
  const subject = `Portfolio: new contact from ${a.name}`;
  const body =
    `From: ${a.name} <${a.email}>\n` +
    `When: ${a.createdAt}\n` +
    `IP:   ${a.ip}\n` +
    `\n${a.message}\n`;

  await ses.send(
    new SendEmailCommand({
      Source: a.from,
      Destination: { ToAddresses: [a.to] },
      ReplyToAddresses: [a.replyTo],
      Message: {
        Subject: { Data: subject },
        Body: { Text: { Data: body } },
      },
    }),
  );
}
